### ISSUE TITLE
Muitli-processing continues outside of pl.Trainer

### ISSUE THREAD

**Issue by Rizhiy** [PyTorchLightning/pytorch-lightning#7699]
## 🐛 Bug

Code which is outside pl.Trainer is being run in multiple processes.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1EqzPuipufH2Rb1nyiliRLoEknZmNbZJD?usp=sharing

### Expected behavior

I realize that this behaviour might be expected in PL, but it is very confusing.
I expect only code inside trainer to be run in multiple processes, but it appears that all code in the file is run in each process.

### Environment
* CUDA:
        - GPU:
                - GeForce RTX 2080 Ti
                - GeForce RTX 2080 Ti
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.20.3
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1+cu102
        - pytorch-lightning: 1.3.2
        - tqdm:              4.60.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #80-Ubuntu SMP Mon Apr 12 17:35:00 UTC 2021

### Additional context

Basically, the way `ddp` works now, it forces creation of dedicated scripts for each task/wrapping all external logic into `if rank == 0`. This makes it very difficult to use `ddp` in any sort of production, thereby forcing the user to write a separate code for training and production.

A few ideas on how this can potentially be solved:
* Perhaps run some analysis on code in the file and create a copy which removes all code not used by trainer, save this copy somewhere and run it instead.
* Since almost everything can be made pickleable in Python, perhaps pickle all inputs to trainer, create new processes with some default code which is able to receive pickled inputs and run them instead.

**Comment by Rizhiy** [comment]
Probably related to #7638

**Comment by tchaton** [comment]
Hey @Rizhiy,

This is an expected behaviour. DDP creates subprocesses meaning the entire script is going to run.
You can use `ddp_spawn` which should solve your problems :)

Best,
T.C

**Comment by kaushikb11** [comment]
@Rizhiy Closing this issue, as it is expected behavior. Feel free to reopen it if you face any issues.

**Comment by Rizhiy** [comment]
@tchaton Ok, thanks. I took a look at `ddp_spawn` and I don't quite understand the second limitation:
> Dataloader(num_workers=N), where N is large, bottlenecks training with DDP… ie: it will be VERY slow or won’t work at all. This is a PyTorch limitation.
What should I set `num_workers` to in `ddp_spawn`. In my PyTorch code I use one worker per process and it works fine with 8 GPUs.
